cffi_builders/lib/ract_dma.c

```diff
@@ -3,40 +3,144 @@
 #include <unistd.h>
 //#include "ract_dma.h"
 
-void set_pin(int value);
+// RPI references
+#define GSCLK           RPI_GPIO_P1_03
+#define DCPRG           RPI_GPIO_P1_05
+#define SCLK            RPI_GPIO_P1_08
+#define XLAT            RPI_GPIO_P1_10
+#define BLANK           RPI_GPIO_P1_11
+#define SIN             RPI_GPIO_P1_12
+#define VPRG            RPI_GPIO_P1_13
+#define OUT             BCM2835_GPIO_FSEL_OUTP
+// TLC5940 references
+#define DC_LENGTH       6
+#define GS_LENGTH       12
+#define GS_CLOCK_CYCLES 4096
+// Internal references
+#define TRUE            0x1
+#define FALSE           0x0
 
-//#define PIN   RPI_GPIO_P1_15
+// Local global variables
+uint8_t     u8_first_gs_cycle;
 
-void initializeGpioInterface(void) {
-    bcm2835_init();
+// Initialize the BCM
+int initialize_RPI_GPIO(void) {
+    int i_ret_val = bcm2835_init();
+    printf("RPI GPIO interface initialized\n");
+    return i_ret_val;
 }
 
-void setup_pin(void) {
-    printf("Setting pin %d as output\n", RPI_GPIO_P1_15);
-    bcm2835_gpio_fsel(RPI_GPIO_P1_15, BCM2835_GPIO_FSEL_OUTP);
-    printf("Set pin %d as output\n", RPI_GPIO_P1_15);
+int close_RPI_GPIO(void) {
+    int i_ret_val = bcm2835_close();
+    printf("RPI GPIO interface closed\n");
+    return i_ret_val;
 }
 
-void set_pin(int value) {
-	if (value == 1) {
-	    printf("Setting pin ON\n");
-		bcm2835_gpio_write(RPI_GPIO_P1_15, HIGH);
-	    printf("Set pin ON\n");
-	} else {
-	    printf("Setting pin OFF\n");
-		bcm2835_gpio_write(RPI_GPIO_P1_15, LOW);
-	    printf("Set pin OFF\n");
-	}
+// Set initial pin directions and initial values
+void setup_interface_pins(void) {
+    bcm2835_gpio_fsel(GSCLK, OUT);
+    bcm2835_gpio_fsel(DCPRG, OUT);
+    bcm2835_gpio_fsel(VPRG, OUT);
+    bcm2835_gpio_fsel(SCLK, OUT);
+    bcm2835_gpio_fsel(XLAT, OUT);
+    bcm2835_gpio_fsel(BLANK, OUT);
+    bcm2835_gpio_fsel(SIN, OUT);
+    set_pin(GSCLK, LOW);
+    set_pin(DCPRG, LOW);
+    set_pin(VPRG, HIGH);
+    set_pin(SCLK, LOW);
+    set_pin(XLAT, LOW);
+    set_pin(BLANK, HIGH);
+    u8_first_gs_cycle = TRUE;
+    printf("Interface configured\n");
 }
 
-void setPinByArray(uint8_t * intArray, int lenArray) {
+// Compensate for board protection logic flip
+void set_pin(uint8_t u8_pin, uint8_t u8_level) {
+    bcm2835_gpio_write(u8_pin, (u8_level & HIGH) ? LOW : HIGH);
+    bcm2835_gpio_write(u8_pin, (u8_level & HIGH) ? LOW : HIGH);
```

Comment:
Do we really want `u8_level & HIGH` rather than just `u8_level`? I'd be pretty surprised if, e.g. `2` was treated the same as `0` (as it is with the current code).